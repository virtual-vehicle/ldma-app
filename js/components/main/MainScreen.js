import React, { Component } from 'react';
import {
  View,
  TouchableOpacity,
  Text,
  StyleSheet,
  Image,
  Dimensions,
} from 'react-native';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import SideMenu from 'react-native-side-menu';
import PercentageCircle from 'react-native-percentage-circle';
import { LineChart, Grid } from 'react-native-svg-charts';
import { COLORS } from '../../constants/colors';
import { GO_TO_SAFE_DRIVING } from '../../actions/actionTypes';
import {
} from '../../actions/uiActions';
import Loader from '../common/Loader';
import Menu from '../common/Menu';
import { getTripsInfo } from '../../actions/tripsInfoActions';

/* Config/Constants
============================================================================= */

/* eslint-disable global-require */

const FIELDS_WIDTH = '88%';
const BUTTON_HEIGHT = 50;
const SCREEN_WIDTH = Dimensions.get('window').width;

type Props = {
  dispatch: any,
  navigation: any,
  login: any,
  loading: boolean,
};

export class MainScreen extends Component<Props, State> {
  constructor(props) {
    super(props);

    this.toggle = this.toggle.bind(this);

    this.state = {
      isOpen: false,
    };
  }

  onMenuItemSelected = () =>
    this.setState({
      isOpen: false,
    });

  toggle() {
    this.setState({
      isOpen: !this.state.isOpen,
    });
  }

  updateMenuState(isOpen) {
    this.setState({ isOpen });
  }

  render() {
    const { isOpen } = this.state;
    const { navigation, loading } = this.props;
    const menu = <Menu onItemSelected={this.onMenuItemSelected} navigation={navigation} />;
    const data = [50, 10, 40, 95, -4, -24, 85, 91, 35, 53, -53, 24, 50, -20, -80];

    const contentInset = { top: 20, bottom: 20 };

    return (
      <SideMenu
        menu={menu}
        isOpen={isOpen}
        onChange={isOpen => this.updateMenuState(isOpen)}
      >
        <View style={styles.container}>
          <TouchableOpacity
            onPress={this.toggle}
            style={styles.menuButton}
          >
            <Image
              source={require('./img/menu.png')}
              style={styles.menu}
            />
          </TouchableOpacity>
          <View style={styles.header}>
            <Text style={styles.headerText}>TripViz</Text>
          </View>
          <View style={styles.content}>
            <View style={styles.percent}>
              <PercentageCircle
                radius={67}
                percent={50}
                color={COLORS.GREEN4}
                borderWidth={10}
                textStyle={{ fontSize: 30 }}
              />
              <Text style={{ paddingTop: 10 }}>Driving Score (total)</Text>
            </View>
            <View>
              <View style={{ display: 'flex', flexDirection: 'row' }}>
                <Text style={styles.cube}>{'12\ntrips'}</Text>
                <Text style={styles.cube}>{'365\nkm'}</Text>
              </View>
              <View style={{ display: 'flex', flexDirection: 'row', marginTop: 10 }}>
                <Text style={styles.cube}>{'13\nevents'}</Text>
                <Text style={styles.cube}>{'13:21\nmin'}</Text>
              </View>
            </View>
          </View>
          <View>
            <LineChart
              style={{ height: 350, width: SCREEN_WIDTH * .9 }}
              data={ data }
              svg={{ stroke: COLORS.GREEN4 }}
              contentInset={ contentInset }
            >
              <Grid />
            </LineChart>
          </View>
          <TouchableOpacity
            style={styles.goToNextScreen}
            onPress={() => navigation.dispatch({ type: GO_TO_SAFE_DRIVING })}
          >
            <Text style={styles.goToNextScreenText}>{`Safe Driving`}</Text>
          </TouchableOpacity>
          {loading && <Loader />}
        </View>
      </SideMenu>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: COLORS.WHITE,
    alignItems: 'center',
  },
  content: {
    display: 'flex',
    flexDirection: 'row',
    marginTop: 40,
  },
  cube: {
    borderColor: COLORS.GREEN4,
    borderWidth: 3,
    fontSize: 16,
    padding: 2,
    paddingTop: 5,
    width: 90,
    textAlign: 'center',
    marginLeft: 10,
  },
  percent: {
    display: 'flex',
    alignItems: 'center',
    marginRight: 20,
  },
  menuButton: {
    position: 'absolute',
    top: 8,
    left: 15,
    padding: 10,
    zIndex: 1000,
  },
  menu: {
    width: 30,
    height: 30,
  },
  header: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: 60,
    backgroundColor: COLORS.BLUE,
  },
  headerText: {
    color: COLORS.WHITE,
    fontSize: 20,
  },
  goToNextScreen: {
    position: 'absolute',
    bottom: SCREEN_WIDTH * 0.10,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: COLORS.BLUE,
    width: FIELDS_WIDTH,
    borderRadius: 5,
    height: BUTTON_HEIGHT,
  },
  goToNextScreenText: {
    color: COLORS.WHITE,
    fontSize: 20,
  },
});

const mapStateToProps = (state) =>
  ({
    loading: state.loading,
    user: state.auth.user,
  });

const mapDispatchToProps = (dispatch) =>
  bindActionCreators({
    getTripsInfo,
  }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(MainScreen);
