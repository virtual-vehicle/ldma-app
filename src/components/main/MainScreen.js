import React, { Component } from 'react';
import {
  View,
  TouchableOpacity,
  Text,
  StyleSheet,
  Image,
  Dimensions,
} from 'react-native';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { get as safeGet } from 'lodash';
import SideMenu from 'react-native-side-menu';
import PercentageCircle from 'react-native-percentage-circle';
import { LineChart, Grid, YAxis, XAxis } from 'react-native-svg-charts';
import { COLORS } from 'ldmaapp/src/constants/colors';
import {
} from 'ldmaapp/src/actions/uiActions';
import Loader from 'ldmaapp/src/components/common/Loader';
import Menu from 'ldmaapp/src/components/common/Menu';
import { getTripsInfo } from 'ldmaapp/src/actions/tripsInfoActions';
import { getGraphTripscore } from 'ldmaapp/src/actions/graphTripscoreActions';
import NavigationService from 'ldmaapp/src/utils/navigation';
import { convertMinutesToHoursMinutes } from 'ldmaapp/src/utils/format';
/* Config/Constants
============================================================================= */

/* eslint-disable global-require */

const FIELDS_WIDTH = '88%';
const BUTTON_HEIGHT = 50;
const SCREEN_WIDTH = Dimensions.get('window').width;

type Props = {
  dispatch: any,
  login: any,
  loading: boolean,
  tripsInfo: Array,
  graphTripscoreList: Array,
};

export class MainScreen extends Component<Props, State> {
  constructor(props) {
    super(props);

    this.toggle = this.toggle.bind(this);

    this.state = {
      isOpen: false,
    };
  }

  componentDidMount() {
    const { getTripsInfo, getGraphTripscore, user } = this.props;
    const auth_token = safeGet(user, 'auth_token', '');
    getTripsInfo(auth_token);
    getGraphTripscore(auth_token);
  }

  onMenuItemSelected = () =>
    this.setState({
      isOpen: false,
    });

  toggle() {
    this.setState({
      isOpen: !this.state.isOpen,
    });
  }

  updateMenuState(isOpen) {
    this.setState({ isOpen });
  }

  render() {
    const { isOpen } = this.state;

    const { loading, tripsInfo, graphTripscoreList } = this.props;
    const menu = <Menu onItemSelected={this.onMenuItemSelected} />;

    let data = [];
    if (graphTripscoreList) {
      data = graphTripscoreList.map(scoreObj => scoreObj.risk_score);
    }

    const numTrips = safeGet(tripsInfo, 'tripsList[0].value', '');
    const numEvents = safeGet(tripsInfo, 'tripsList[1].value', '');
    const distance = safeGet(tripsInfo, 'tripsList[2].value', '');
    const totalTime = safeGet(tripsInfo, 'tripsList[3].value', '');
    const driverScore = safeGet(tripsInfo, 'tripsList[7].value', '');

    const contentInset = { top: 20, bottom: 20 };

    let startDate = '';
    if (graphTripscoreList) {
      startDate = safeGet(graphTripscoreList[0], 'start_at', '');
    }

    let endDate = '';
    if (graphTripscoreList) {
      const index = graphTripscoreList.length -1;
      endDate = safeGet(graphTripscoreList[index], 'start_at', '');
    }

    return (
      <SideMenu
        menu={menu}
        isOpen={isOpen}
        onChange={isOpen => this.updateMenuState(isOpen)}
      >
        <View style={styles.container}>
          <TouchableOpacity
            onPress={this.toggle}
            style={styles.menuButton}
          >
            <Image
              source={require('ldmaapp/assets/png/menu.png')}
              style={styles.menu}
            />
          </TouchableOpacity>
          <View style={styles.header}>
            <Text style={styles.headerText}>TripViz</Text>
          </View>
          <View style={styles.content}>
            <View style={styles.percent}>
              <PercentageCircle
                radius={67}
                percent={driverScore}
                color={COLORS.GREEN4}
                borderWidth={10}
                textStyle={{ fontSize: 30 }}
              />
              <Text style={{ paddingTop: 10 }}>Driving Score (total)</Text>
            </View>
            <View>
              <View style={{ display: 'flex', flexDirection: 'row' }}>
                <Text style={styles.cube}>{`${numTrips}\ntrips`}</Text>
                <Text style={styles.cube}>{`${distance}\nkm`}</Text>
              </View>
              <View style={{ display: 'flex', flexDirection: 'row', marginTop: 10 }}>
                <Text style={styles.cube}>{`${numEvents}\nevents`}</Text>
                <Text style={styles.cube}>{convertMinutesToHoursMinutes(totalTime)}</Text>
              </View>
            </View>
          </View>
          <Text style={{ marginTop: 20 }}>Last 20 trip scores</Text>
          <View style={{ flexDirection: 'row' }}>
            <YAxis
              data={data}
              contentInset={contentInset}
              svg={{
                fill: 'grey',
                fontSize: 10,
              }}
              numberOfTicks={ 10 }
              formatLabel={ value => value }
            />
            <View>
              <LineChart
                style={{ height: 250, width: SCREEN_WIDTH * .9 }}
                data={ data }
                svg={{ stroke: COLORS.GREEN4 }}
                contentInset={ contentInset }
              >
              <Grid />
              </LineChart>
              <XAxis
                style={{ marginHorizontal: -10 }}
                data={ data }
                formatLabel={ (value, index) => {
                  if (index === 0) {
                    return startDate;
                  }
                  else if (index === 19) {
                    return endDate;
                  }
                }}
                contentInset={{ left: 10, right: 10 }}
                svg={{ fontSize: 10, fill: 'black' }}
              />
            </View>
          </View>
          <TouchableOpacity
            style={styles.goToNextScreen}
            onPress={() => NavigationService.navigate('SafeDriving')}
          >
            <Text style={styles.goToNextScreenText}>{`Safe Driving`}</Text>
          </TouchableOpacity>
          {loading && <Loader />}
        </View>
      </SideMenu>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: COLORS.WHITE,
    alignItems: 'center',
  },
  content: {
    display: 'flex',
    flexDirection: 'row',
    marginTop: 40,
  },
  cube: {
    borderColor: COLORS.GREEN4,
    borderWidth: 3,
    fontSize: 16,
    padding: 2,
    paddingTop: 5,
    width: 90,
    textAlign: 'center',
    marginLeft: 10,
  },
  percent: {
    display: 'flex',
    alignItems: 'center',
    marginRight: 20,
  },
  menuButton: {
    position: 'absolute',
    top: 8,
    left: 15,
    padding: 10,
    zIndex: 1000,
  },
  menu: {
    width: 30,
    height: 30,
  },
  header: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: 60,
    backgroundColor: COLORS.BLUE,
  },
  headerText: {
    color: COLORS.WHITE,
    fontSize: 20,
  },
  goToNextScreen: {
    position: 'absolute',
    bottom: SCREEN_WIDTH * 0.10,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: COLORS.BLUE,
    width: FIELDS_WIDTH,
    borderRadius: 5,
    height: BUTTON_HEIGHT,
  },
  goToNextScreenText: {
    color: COLORS.WHITE,
    fontSize: 20,
  },
});

const mapStateToProps = (state) =>
  ({
    loading: state.loading,
    user: state.auth.user,
    tripsInfo: state.tripsInfo,
    graphTripscoreList: state.graphTripscore.tripscoreList,
  });

const mapDispatchToProps = (dispatch) =>
  bindActionCreators({
    getTripsInfo,
    getGraphTripscore,
  }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(MainScreen);
