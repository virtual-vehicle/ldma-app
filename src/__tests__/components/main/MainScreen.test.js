import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { MainScreen } from 'ldmaapp/src/components/main/MainScreen';

Enzyme.configure({ adapter: new Adapter() });

function setup() {
  const props = {
    getTripsInfo: jest.fn(),
  };

  const wrapper = shallow(<MainScreen {...props} />);

  return {
    props,
    wrapper,
  };
}

describe('MainScreen component', () => {
  it('should renders correctly', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
