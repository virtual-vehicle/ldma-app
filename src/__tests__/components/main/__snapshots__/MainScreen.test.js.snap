// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen component should renders correctly 1`] = `
<SideMenu
  animationFunction={[Function]}
  animationStyle={[Function]}
  autoClosing={true}
  bounceBackOnOverdraw={true}
  disableGestures={false}
  edgeHitWidth={60}
  hiddenMenuOffset={0}
  isOpen={false}
  menu={
    <Connect(Menu)
      onItemSelected={[Function]}
    />
  }
  menuPosition="left"
  onChange={[Function]}
  onMove={[Function]}
  onSliding={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  openMenuOffset={500}
  toleranceX={10}
  toleranceY={10}
>
  <ImageBackground
    source={null}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFF",
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Object {
          "left": 15,
          "padding": 10,
          "position": "absolute",
          "top": 8,
          "zIndex": 1000,
        }
      }
    >
      <Image
        source={null}
        style={
          Object {
            "height": 30,
            "width": 30,
          }
        }
      />
    </TouchableOpacity>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "display": "flex",
          "height": 60,
          "justifyContent": "center",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFF",
            "fontSize": 20,
          }
        }
      >
        TripViz
      </Text>
    </View>
    <View
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "marginTop": 40,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "marginRight": 20,
          }
        }
      >
        <PercentageCircle
          bgcolor="#e3e3e3"
          borderWidth={10}
          innerColor="#fff"
          percent=""
          radius={67}
          textStyle={
            Object {
              "fontSize": 30,
            }
          }
        />
        <Text
          style={
            Object {
              "paddingTop": 10,
            }
          }
        >
          Driving Score (total)
        </Text>
      </View>
      <View>
        <View
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "borderColor": "#32CD32",
                "borderWidth": 3,
                "fontSize": 16,
                "marginLeft": 10,
                "padding": 2,
                "paddingTop": 5,
                "textAlign": "center",
                "width": 90,
              }
            }
          >
            
trips
          </Text>
          <Text
            style={
              Object {
                "borderColor": "#32CD32",
                "borderWidth": 3,
                "fontSize": 16,
                "marginLeft": 10,
                "padding": 2,
                "paddingTop": 5,
                "textAlign": "center",
                "width": 90,
              }
            }
          >
            
km
          </Text>
        </View>
        <View
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
              "marginTop": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "borderColor": "#32CD32",
                "borderWidth": 3,
                "fontSize": 16,
                "marginLeft": 10,
                "padding": 2,
                "paddingTop": 5,
                "textAlign": "center",
                "width": 90,
              }
            }
          >
            
events
          </Text>
          <Text
            style={
              Object {
                "borderColor": "#32CD32",
                "borderWidth": 3,
                "fontSize": 16,
                "marginLeft": 10,
                "padding": 2,
                "paddingTop": 5,
                "textAlign": "center",
                "width": 90,
              }
            }
          >
            0 min
          </Text>
        </View>
      </View>
    </View>
    <Text
      style={
        Object {
          "marginTop": 20,
        }
      }
    >
      Last 20 trip scores
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <YAxis
        contentInset={
          Object {
            "bottom": 20,
            "top": 20,
          }
        }
        data={Array []}
        formatLabel={[Function]}
        numberOfTicks={10}
        scale={[Function]}
        spacingInner={0.05}
        spacingOuter={0.05}
        svg={
          Object {
            "fill": "grey",
            "fontSize": 10,
          }
        }
        yAccessor={[Function]}
      />
      <View>
        <LineChart
          contentInset={
            Object {
              "bottom": 20,
              "top": 20,
            }
          }
          curve={[Function]}
          data={Array []}
          height={100}
          numberOfTicks={10}
          style={
            Object {
              "height": 250,
              "width": 675,
            }
          }
          svg={
            Object {
              "stroke": undefined,
            }
          }
          width={100}
          xAccessor={[Function]}
          xScale={[Function]}
          yAccessor={[Function]}
          yScale={[Function]}
        >
          <Grid
            belowChart={true}
            direction="HORIZONTAL"
          />
        </LineChart>
        <XAxis
          contentInset={
            Object {
              "left": 10,
              "right": 10,
            }
          }
          data={Array []}
          formatLabel={[Function]}
          scale={[Function]}
          spacingInner={0.05}
          spacingOuter={0.05}
          style={
            Object {
              "marginHorizontal": -10,
            }
          }
          svg={
            Object {
              "fill": "black",
              "fontSize": 10,
            }
          }
          xAccessor={[Function]}
        />
      </View>
    </View>
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "rgba(255, 255, 255, 0.5)",
          "borderRadius": 50,
          "borderWidth": 1,
          "height": 50,
          "justifyContent": "center",
          "marginTop": 40,
          "width": 630,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFF",
            "fontSize": 20,
          }
        }
      >
        Safe Driving
      </Text>
    </TouchableOpacity>
  </ImageBackground>
</SideMenu>
`;
